### DISCUSSION TITLE
Schema class api suggested additions

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#151] [Design & Development]
This came up because I made a unilateral decision to start writing a set of standard test functions that would be run through a standalone program we very much need to have that I recommend we call either dbverify or just verify.  The purpose of dbverify is to run one or more tests on the db to provide a guaranteed that readers in a job won't abort.   I think with that tool we can make "pedantic" the default for reading.  We can work out more details about that later, but in playing with some prototypes I quickly recognized some holes I think exist in the api of the Schema class.  They are:
1. I think we need to extend the class to have a cleaner support for normalized data.   This will require a change to yaml file for sure and some additions to the class definition to go with them. Some things I suggest as a starting point are:

- **normalized**  return a list of collection names that are used for normalization
- **normalization_id** return the key used normalizing with a given collection name.  This would regularize the way we do this create names like 'site_id' and 'source_id'.   
- **normalized_key** standardize the way a normalization key would be created.  Suggest this signature:  `def normalized_key(collection,key):` where collection is the name of the collection containing key.  With the current model, for example, if collection='site' and key='lat' this would return 'site_lat'.   This would be like normalization_id but for regular attributes.

2.  The Schema class itself needs a clean way to ask if a collection name is defined in the schema.  Right now the only way I can see to validate if a collection name is know is to let Schema throw an exception when you try to use an undefined name.  Let me clarify that with a small fragment of code:
```
db=Database(client,'testdb')
# This method returns the default schema defined for this Database instance
dbschema=db.database_schema
# this is the construct to retrieve the schema definition for "collection"
col_schema=dbschema[collection]
```
The problem right now is the call to dbschema[collection] will throw an exception if collection is not a valid name for the schema.  That is a the right behavior, but in many cases it would be preferable to test of an illegal value without depending on an exception mechanism handle the problem.   Hence, I think it would be helpful to have a method in the class dbschema (I think it is DBSchemaDefinition) that can test for validity of a collection name.  Maybe this signature:
```
def collection_defined(self,name): 
    '''
    Return true if name is a valid collection. False if it is not defined in the schema.
    Note for design - this should be a method of DBSchemaDefinition
    """
```
I would have implemented this had I known how.  I think it can be done by looking at whatever dir(classname) returns.   I have no idea how to do that within a python function.

That is all I have for now, but it is likely I will encounter others as I wade through this.

**Comment by pavlis** [comment]
Here is a related suggestion.   Once created a database should have one and only one schema - at least I think so.   Seems to me we could automate that and have a special collection to hold that and other potential common database attributes.  The content could be as simple as the schema name or it could contain something as complete as a copy of the data in a Schema object derived from the yaml file.  I see a thing called "schema_dic" but there seem to be several other pieces of data spread through the different classes in schema.py.    Certainly not a high priority to improve performance since the schema classes tend to be create as one of the first lines in a workflow and then only be copied.   

A broader issue is if there are other things that might belong in something like a "global" collection?   I think on old version of the docs penciled in the idea, but we as yet don't have any such thing.  Might be a place for global history data, but we haven't spent any time thinking about that recently.

**Reply by haruming** [reply]
I agree that once created a database should have one and only one schema, but I am a little confused about your suggestion. In our current implementation, once we create a database, we can pass the db_schema and md_schema for initialization and the schemas are fixed for that database if I understand correctly. So why do we need to have a special collection to hold that and other potential common database attributes?

**Comment by pavlis** [comment]
It is quite possible I don't understand all the intricate details of the set of classes in the schema.py file.   It is a reason (hint hint, Ian) we need more complete documentation on the what the schema capability is supposed to work.

To clarify what I meant was if an application opens an existing database it should be able to automatically infer the schema if it was created by MsPASS.  If the database doesn't have a schema defined somehow the application can test it should abort.   That would be a basic sanity check to make a clear error condition instead of depending upon something else to abort downstream.   The sanity check would be a check against (a) a typo for the name, and (b) pointing at a wrong (valid) database.  The later could happen easily, for example,  if a user had multiple datasets defined with different schemas.

**Reply by wangyinz** [reply]
hmmm, it seems we will have to separate database schema and metadata schema from now on. The suggestion you made here really only applies to the database schema, because the user will want to use different metadata schema to read the same database. I guess we should just take apart the `mspass.yaml` file into `mspass_db.yaml` and `mspass_md.yaml`. We will also need to change the constructor of the database class such that it reads the metadata schema from the yaml file and gets the database schema from the database. Well, what if the user creates a new database that has no schema? We probably should still allow reading in database schema from a file. I guess the question is how to deal with inconsistent database schema from database and yaml file?

**Reply by pavlis** [reply]
Well, first what I suggested is not essential.  We can just make it clear that the user is responsible for keeping track of any nonstandard schema.   That may be the best approach for now as we have bigger issues to solve than this one.   Maybe we should shelf this topic until we have time to deal with it.  I think it would be a clean retrofit after the fact.

**Reply by wangyinz** [reply]
I agree

**Comment by wangyinz** [comment]
Back to the original suggestion. I have implemented the `__contains__` method for the `SchemaBase` class (in https://github.com/wangyinz/mspass/commit/eb7c4c02b50832cf0cd0d6e99131ee2865cf7881), so testing whether a collection (or data object for MetadataSchema) is defined can be done with a `'key' in schema` statement.

For the normalized data support, I don't think our current schema class can have those methods defined cleanly. We currently have no notation of normalization in the yaml file, and I am not sure how to add it unambiguously with current file structure. I guess the issue here is that the normalization is something only applies to a metadata schema. However, it also relates to the database schema. Another thing to keep in mind here is that normalization is a relative concept. For example, if we concentrate on the `wf` collection, the keys like `source_lat` is certainly a normalized key. However, we may also put a `wf_Seismogram_id` key to the `source` collection, then the `_id` of `wf_Seismogram` becomes normalized key of the `source`. Then, if you think it carefully, whether a key is normalized or not does not really matter to the metadata because what we really care in the metadata is where in the database we should look up the key for. This is already covered by the `MDSchemaDefinition.collection(key)` method. I guess what's missing is a method to get the aliased `_id` attribute that indexes the `key`. This is currently done with https://github.com/wangyinz/mspass/blob/eb7c4c02b50832cf0cd0d6e99131ee2865cf7881/python/mspasspy/db/database.py#L210
where we implicitly embedded the rule of `col + '_id'` to retrieve the aliased `_id`. I guess we can encapsulate that into a `MDSchemaDefinition.collection_id(key)` method, but it will simply do the same `col + '_id'` concatenation there.

**Reply by wangyinz** [reply]
Probably `collection_id` is too confusing. Maybe we should name that method `reference_id`? I don't want to call it `normalization_id` as I think normalization is also a confusing concept here as discussed above.
